File: python/src/raidxai/resources/video.py
"""Video forensics — detect deepfake / AI-generated video. Asynchronous: submit, then poll."""

from __future__ import annotations

from collections.abc import Callable
from typing import TYPE_CHECKING

from .. import _specs
from .._poll import poll_until, poll_until_async
from .._specs import PagedResult, normalize_paged
from ..files import FileInput
from ..models import TERMINAL_JOB_STATUSES, VideoForensicsCreditInfo, VideoJob, VideoSubmitResponse

if TYPE_CHECKING:
    from ..aio import AsyncRaidClient
    from ..client import RaidClient


def _page(raw: object) -> PagedResult[VideoJob]:
    items, total = normalize_paged(raw)
    return PagedResult[VideoJob](
        items=[VideoJob.model_validate(i) for i in items], total_count=total
    )


def _is_done(job: VideoJob) -> bool:
    return job.status in TERMINAL_JOB_STATUSES


class Video:
    def __init__(self, client: RaidClient) -> None:
        self._client = client

    def submit(
        self,
        file: FileInput,
        *,
        client_duration_seconds: float | None = None,
        source_url: str | None = None,
    ) -> VideoSubmitResponse:
        """Submit a video file. Returns a ``job_id`` to poll. Requires the ``video`` scope."""
        spec = _specs.video_submit(file, client_duration_seconds, source_url)
        return self._client.request(spec, VideoSubmitResponse)

    def submit_from_url(
        self, url: str, *, client_duration_seconds: float | None = None
    ) -> VideoSubmitResponse:
        """Submit a video from a public URL. Returns a ``job_id`` to poll."""
        spec = _specs.video_submit_from_url(url, client_duration_seconds)
        return self._client.request(spec, VideoSubmitResponse)

    def get_job(self, job_id: str) -> VideoJob:
        """Fetch a video job. ``result`` is populated once ``status`` is ``Completed``."""
        return self._client.request(_specs.video_get_job(job_id), VideoJob)

    def list_jobs(
        self, *, skip: int | None = None, take: int | None = None
    ) -> PagedResult[VideoJob]:
        """List submitted video jobs, most recent first."""
        return _page(self._client.request(_specs.video_list_jobs(skip, take)))

    def cancel(self, job_id: str) -> None:
        """Cancel a running video job."""
        self._client.request(_specs.video_cancel(job_id))

    def submit_and_wait(
        self,
        file: FileInput,
        *,
        client_duration_seconds: float | None = None,
        source_url: str | None = None,
        interval_seconds: float = 3.0,
        timeout_seconds: float = 300.0,
        on_poll: Callable[[VideoJob], None] | None = None,
    ) -> VideoJob:
        """Submit a video and poll until it reaches a terminal state."""
        submitted = self.submit(
            file, client_duration_seconds=client_duration_seconds, source_url=source_url
        )
        return poll_until(
            lambda: self.get_job(submitted.job_id),
            _is_done,
            lambda j: j.status.value,
            interval_seconds=interval_seconds,
            timeout_seconds=timeout_seconds,
            on_poll=on_poll,
        )

    def credit_info(self) -> VideoForensicsCreditInfo:
        """Read the credit balance and the live video limits: cost per frame, frames analyzed per
        second, the longest video allowed and what it costs, max file size, allowed formats, and
        today's usage against the daily cap.

        Read these rather than hard-coding the defaults — they are account settings and can
        change without an SDK release. Any valid API token may call this; no scope is required.
        """
        return self._client.request(_specs.video_credit_info(), VideoForensicsCreditInfo)


class AsyncVideo:
    def __init__(self, client: AsyncRaidClient) -> None:
        self._client = client

    async def submit(
        self,
        file: FileInput,
        *,
        client_duration_seconds: float | None = None,
        source_url: str | None = None,
    ) -> VideoSubmitResponse:
        spec = _specs.video_submit(file, client_duration_seconds, source_url)
        return await self._client.request(spec, VideoSubmitResponse)

    async def submit_from_url(
        self, url: str, *, client_duration_seconds: float | None = None
    ) -> VideoSubmitResponse:
        spec = _specs.video_submit_from_url(url, client_duration_seconds)
        return await self._client.request(spec, VideoSubmitResponse)

    async def get_job(self, job_id: str) -> VideoJob:
        return await self._client.request(_specs.video_get_job(job_id), VideoJob)

    async def list_jobs(
        self, *, skip: int | None = None, take: int | None = None
    ) -> PagedResult[VideoJob]:
        return _page(await self._client.request(_specs.video_list_jobs(skip, take)))

    async def cancel(self, job_id: str) -> None:
        await self._client.request(_specs.video_cancel(job_id))

    async def submit_and_wait(
        self,
        file: FileInput,
        *,
        client_duration_seconds: float | None = None,
        source_url: str | None = None,
        interval_seconds: float = 3.0,
        timeout_seconds: float = 300.0,
        on_poll: Callable[[VideoJob], None] | None = None,
    ) -> VideoJob:
        submitted = await self.submit(
            file, client_duration_seconds=client_duration_seconds, source_url=source_url
        )
        return await poll_until_async(
            lambda: self.get_job(submitted.job_id),
            _is_done,
            lambda j: j.status.value,
            interval_seconds=interval_seconds,
            timeout_seconds=timeout_seconds,
            on_poll=on_poll,
        )

    async def credit_info(self) -> VideoForensicsCreditInfo:
        return await self._client.request(_specs.video_credit_info(), VideoForensicsCreditInfo)

File: typescript/src/resources/video.ts
import type { RaidClient } from "../client.js";
import type {
  PagedResult,
  VideoForensicsCreditInfo,
  VideoJob,
  VideoSubmitResponse,
} from "../types.js";
import { TERMINAL_JOB_STATUSES } from "../types.js";
import { appendFile, type FileInput } from "../upload.js";
import { pollUntil, type PollOptions } from "../poll.js";
import { normalizePaged } from "./paging.js";

export interface VideoSubmitOptions {
  /**
   * Video duration in seconds, read from the file by your client. Strongly recommended
   * so credits are reserved accurately at submit time.
   */
  clientDurationSeconds?: number;
  /** Source URL of the video, for your own auditing. */
  sourceUrl?: string;
}

export interface VideoSubmitFromUrlOptions {
  clientDurationSeconds?: number;
}

export interface VideoListOptions {
  skip?: number;
  take?: number;
}

/** Video forensics — detect deepfake / AI-generated video. Asynchronous: submit, then poll. */
export class Video {
  constructor(private readonly client: RaidClient) {}

  /** Submit a video file for analysis. Returns a `jobId` to poll. Requires the `video` scope. */
  submit(file: FileInput, options: VideoSubmitOptions = {}): Promise<VideoSubmitResponse> {
    const form = new FormData();
    appendFile(form, "file", file);
    if (options.clientDurationSeconds !== undefined) {
      form.append("clientDurationSeconds", String(options.clientDurationSeconds));
    }
    if (options.sourceUrl !== undefined) form.append("sourceUrl", options.sourceUrl);

    return this.client.request<VideoSubmitResponse>({
      method: "POST",
      path: "/api/app/video-forensics/submit",
      form,
    });
  }

  /** Submit a video fetched from a public URL. Returns a `jobId` to poll. */
  submitFromUrl(url: string, options: VideoSubmitFromUrlOptions = {}): Promise<VideoSubmitResponse> {
    return this.client.request<VideoSubmitResponse>({
      method: "POST",
      path: "/api/app/video-forensics/submit-from-url",
      json: { url, clientDurationSeconds: options.clientDurationSeconds },
    });
  }

  /** Fetch the current state of a video job. `result` is populated once `status` is `Completed`. */
  getJob(jobId: string): Promise<VideoJob> {
    return this.client.request<VideoJob>({
      method: "GET",
      path: `/api/app/video-forensics/jobs/${encodeURIComponent(jobId)}`,
    });
  }

  /** List submitted video jobs, most recent first. */
  async listJobs(options: VideoListOptions = {}): Promise<PagedResult<VideoJob>> {
    const raw = await this.client.request<unknown>({
      method: "GET",
      path: "/api/app/video-forensics/jobs",
      query: { skip: options.skip, take: options.take },
    });
    return normalizePaged<VideoJob>(raw);
  }

  /** Cancel a running video job. */
  cancel(jobId: string): Promise<void> {
    return this.client.request<void>({
      method: "POST",
      path: `/api/app/video-forensics/jobs/${encodeURIComponent(jobId)}/cancel`,
    });
  }

  /**
   * Submit a video and poll until the job reaches a terminal state
   * (`Completed` / `Failed` / `Cancelled`). Convenience over `submit` + `getJob`.
   */
  async submitAndWait(
    file: FileInput,
    options: VideoSubmitOptions & PollOptions = {},
  ): Promise<VideoJob> {
    const { intervalMs, timeoutMs, onPoll, ...submitOptions } = options;
    const { jobId } = await this.submit(file, submitOptions);
    return pollUntil<VideoJob>(
      () => this.getJob(jobId),
      (job) => TERMINAL_JOB_STATUSES.includes(job.status),
      (job) => job.status,
      { intervalMs, timeoutMs, onPoll: onPoll as PollOptions["onPoll"] },
    );
  }

  /**
   * Read the credit balance and the live video limits: cost per frame, frames analyzed per
   * second, the longest video allowed and what it costs, max file size, allowed formats, and
   * today's usage against the daily cap. Read these rather than hard-coding the defaults — they
   * are account settings and can change without an SDK release. Any valid API token may call
   * this; no scope is required.
   */
  creditInfo(): Promise<VideoForensicsCreditInfo> {
    return this.client.request<VideoForensicsCreditInfo>({
      method: "GET",
      path: "/api/app/video-forensics/credit-info",
    });
  }
}
